devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,182 @@
+
+
+class Differentiable(sympy.Expr):
+    """
+    This class represents Devito differentiable objects such as
+    sum of functions, product of function or FD approximation and
+    provides FD shortcuts for such expressions
+    """
+    _op_priority = 100.0
+
+    def __new__(cls, *args, **kwargs):
+        if cls == Differentiable:
+            assert len(args) == 1
+            expr = args[0]
+            new_obj = sympy.Expr.__new__(cls, expr)
+            # Initialization
+            new_obj.__init__(expr, **kwargs)
+            return new_obj
+        return sympy.Expr.__new__(cls, *args)
+
+    def __init__(self, expr, **kwargs):
+        # Recover the list of possible FD shortcuts
+        self._dtype = kwargs.get('dtype')
+        self._space_order = kwargs.get('space_order')
+        self._time_order = kwargs.get('time_order', 0)
+        self._indices = kwargs.get('indices', ())
+        self._staggered = kwargs.get('staggered')
+        self._grid = kwargs.get('grid')
+        # Generate FD shortcuts for expression or copy from input
+        self._fd = kwargs.get('fd', {})
+        # Save kwargs
+        self._expr = expr
+
+    @property
+    def space_order(self):
+        return self._space_order
+
+    @property
+    def time_order(self):
+        return self._time_order
+
+    @property
+    def staggered(self):
+        return self._staggered
+
+    @property
+    def indices(self):
+        return self._indices
+
+    @property
+    def fd(self):
+        return self._fd
+
+    @property
+    def grid(self):
+        return self._grid
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    def __getattr__(self, name):
+        if name == 'fd' or name == '_fd':
+            raise AttributeError()
+        if name in self.fd.keys():
+            return self.fd[name][0](self)
+        return self.__getattribute__(name)
+
+    def xreplace(self, rule):
+        out = getattr(self, '_expr', self)
+        if out in rule:
+            return rule[out]
+        elif rule:
+            args = []
+            for a in out.args:
+                try:
+                    args.append(a.xreplace(rule))
+                except AttributeError:
+                    args.append(a)
+            args = tuple(args)
+            if not _aresame(args, out.args):
+                return out.func(*args)
+        return out
+
+    def _merge_fd_properties(self, other):
+        merged = {}
+        merged["space_order"] = np.min([getattr(self, 'space_order', 100) or 100,
+                                        getattr(other, 'space_order', 100)])
+        merged["time_order"] = np.min([getattr(self, 'time_order', 100) or 100,
+                                       getattr(other, 'time_order', 100)])
+        merged["indices"] = tuple(set(self.indices + getattr(other, 'indices', ())))
+        merged["fd"] = dict(getattr(self, 'fd', {}), **getattr(other, 'fd', {}))
+        merged["staggered"] = self.staggered
```

Comment:
should we not check that they are actually compatible? 
that's why I was wondering about the use of the word "merge" .